### PR TITLE
Use GL context to draw in a thread save way

### PR DESCRIPTION
**Pull request by PhilippvK**
Feel free to ask me anything if you have questions. Spent multiple days (and nights) figuring this out, getting it work on every operating system (inside Containers).

I probably won't answer today, need some rest first.

### CODE REVIEW

**Comment by PhilippvK** [comment]
Will need some time to rebase this branch to your latest changes

**Comment by alxhoff** [comment]
ok. i'll wait for you to do it and then merge it. What exactly does this do?

**Comment by PhilippvK** [comment]
I will annotate some sourcecode for you when the conflicts are resolved to answer your question.

**Comment by alxhoff** [comment]
Awesome. First a coffee and then i'll check it out. Cheers!

**Comment by PhilippvK** [comment]
No hurries I probably won’t tackle it before lunch. Sorry!

**Comment by alxhoff** [comment]
No worries. Have plenty of other stuff to do.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:171]: A GL Context is required make SDL2 drawing possible among multiple threads on every platform.

Only one thread can own the current context and draw to the window.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: While `SDL_Init ` always has to be called from the Thread 0 running `main()`, at the updates of the screen have to be done in another thread. I solved this by claiming the context and creating the renderer in the second call of the function. I am sure there are other alternatives like having 2 separate functions which are less hacky.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:622]: Just ignore this. (Exporting some environment variables is required on some systems when drawing via X11)

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: `SDL_WINDOW_OPENGL` flag is required to use GL contexts.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: Create Context and check for errors

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: Main Thread claims context and releases it afterwards.

I think I can get rid of both calls as the there is no drawing happening but I left it in to have the ability to create a renderer and add initial draw statements here which will appear before the Sheduler is running.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: `NULL` releases the current context. Now its is free to be claimed by another thread.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: Now the Drawing Thread (`vSwapBuffers`) can claim the context.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: After the context is owned by `vSwapBuffers`, a renderer can be created.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c]: `atexit` only required in first call

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:681]: Proper Cleanup of context.

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [inline lib/Gfx/TUM_Draw.c:690]: You have screwed up the spaces and tabs somehow in a earlier commit. I fixed them here, as I had one indention level more due to the additional if statement.

**Comment by PhilippvK** [comment]
Comments are there, but i still have to test the result. Please wait until done.

**Comment by PhilippvK** [comment]
Yeah the check is failing, I will fix that..

**Comment by PhilippvK** [comment]
I last fix is coming soon...

File: lib/Gfx/TUM_Draw.c
Comment (alxhoff) [inline lib/Gfx/TUM_Draw.c]: `buffer` here should be replaced with `FONT_LOCATION`. I just realized.

**Review by alxhoff — CHANGES_REQUESTED** [review]

**Comment by PhilippvK** [comment]
I think I am finally done.
I can rebase the commit when you are done reviewing, as I do not want to dismiss the comments now.

File: src/main.c
Comment (alxhoff) [inline src/main.c]: Without digging into it, should this be NULL?

**Review by alxhoff — COMMENTED** [review]
Drawing init, should it be passed NULL when called from frame swap?

**Review by PhilippvK — COMMENTED** [review]
> Drawing init, should it be passed NULL when called from frame swap?

It is only used on the first call, so argument is not relevant.

I think having 2 separate functions instead of one would make more sense... (arrrggg)

Do you have recommendations for the naming of the 2nd function used to claim the context and create the renderer?

File: lib/Gfx/TUM_Draw.c
Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c]: Ahhh, makes sense. I think two functions would be a good idea. Maybe something like "bindDrawingToTask"?

File: lib/Gfx/TUM_Draw.c
Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c]: Sounds like a good idea to me. I also have a tendency to leave code that will make future additions easier. Probably bad style but meh.

File: lib/Gfx/TUM_Draw.c
Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c]: Did you find a good site on this context stuff? or is it just in docs? Would be interested in giving it a proper read.

**Review by alxhoff — APPROVED** [review]

**Comment by alxhoff** [comment]
will merge once I am done with my current stuff, working on #20.

**Comment by PhilippvK** [comment]
Its 2 weeks ago since i read about it but i will try to find something

**Comment by PhilippvK** [comment]
Should i restructure everything to be 2 separate function before you merge?

**Comment by alxhoff** [comment]
> Should i restructure everything to be 2 separate function before you merge?

yer not a bad idea!

**Comment by PhilippvK** [comment]
https://discourse.libsdl.org/t/sdl-opengl-context-and-threads/17813/5

**Comment by alxhoff** [comment]
thanks!

**Comment by PhilippvK** [comment]
This should be it for now....

File: lib/Gfx/TUM_Draw.c
Comment (alxhoff) [inline lib/Gfx/TUM_Draw.c:678]: I think the error handling here is a bit mixed up. I will modify it and you can review.

**Review by alxhoff — COMMENTED** [review]
This appears to of fixed #5 :D

File: lib/Gfx/TUM_Draw.c
Comment (PhilippvK) [reply lib/Gfx/TUM_Draw.c:678]: > I think the error handling here is a bit mixed up. I will modify it and you can review.

You probably habe to fix it in both functions. Sorry.

Comment (alxhoff) [reply lib/Gfx/TUM_Draw.c:678]: other one looks good! So glad #5 is fixed. Great work!

**Pull request merged**
